modules/apps/captcha/captcha-impl/src/test/java/com/liferay/captcha/internal/function/captcha/FunctionCaptchaImplTest.java

```diff
@@ -0,0 +1,202 @@
+		);
+
+		Mockito.when(
+			functionCaptchaImplConfiguration.resourcePath()
+		).thenReturn(
+			RandomTestUtil.randomString()
+		);
+
+		FunctionCaptchaImpl functionCaptchaImpl = new FunctionCaptchaImpl();
+
+		ReflectionTestUtil.setFieldValue(
+			functionCaptchaImpl, "_functionCaptchaImplConfiguration",
+			functionCaptchaImplConfiguration);
+
+		Future<byte[]> future = Mockito.mock(Future.class);
+
+		Mockito.when(
+			future.get()
+		).thenReturn(
+			RandomTestUtil.randomBytes()
+		);
+
+		HttpServletRequest httpServletRequest = Mockito.mock(
+			HttpServletRequest.class);
+
+		Mockito.when(
+			httpServletRequest.getRemoteAddr()
+		).thenReturn(
+			RandomTestUtil.randomString()
+		);
+
+		Mockito.when(
+			ParamUtil.getString(
+				httpServletRequest,
+				functionCaptchaImplConfiguration.responseParameterName())
+		).thenReturn(
+			RandomTestUtil.randomString()
+		);
+
+		PortalCatapult portalCatapult = Mockito.mock(PortalCatapult.class);
+
+		Mockito.when(
+			portalCatapult.launch(
+				Mockito.anyLong(), Mockito.any(), Mockito.anyString(),
+				Mockito.any(), Mockito.anyString(), Mockito.anyLong())
+		).thenReturn(
+			future
+		);
+
+		ReflectionTestUtil.setFieldValue(
+			functionCaptchaImpl, "_portalCatapult", portalCatapult);
+
+		User user = Mockito.mock(User.class);
+
+		UserLocalService userLocalService = Mockito.mock(
+			UserLocalService.class);
+
+		Mockito.when(
+			user.getUserId()
+		).thenReturn(
+			RandomTestUtil.randomLong()
+		);
+
+		Mockito.when(
+			userLocalService.getUserByScreenName(
+				Mockito.anyLong(), Mockito.anyString())
+		).thenReturn(
+			user
+		);
+
+		ReflectionTestUtil.setFieldValue(
+			functionCaptchaImpl, "_userLocalService", userLocalService);
+
+		JSONFactory jsonFactory = _mockJSONFactory(true);
+
+		ReflectionTestUtil.setFieldValue(
+			functionCaptchaImpl, "_jsonFactory", jsonFactory);
+
+		functionCaptchaImpl.validateChallenge(httpServletRequest);
+
+		ArgumentCaptor<JSONObject> argumentCaptor = ArgumentCaptor.forClass(
+			JSONObject.class);
+
+		Mockito.verify(
+			portalCatapult, Mockito.times(1)
+		).launch(
+			Mockito.anyLong(), Mockito.any(), Mockito.anyString(),
+			argumentCaptor.capture(), Mockito.anyString(), Mockito.anyLong()
+		);
+
+		JSONObject payloadJSONObject = argumentCaptor.getValue();
+
+		Assert.assertTrue(payloadJSONObject.has("remoteip"));
+		Assert.assertTrue(payloadJSONObject.has("response"));
+
+		jsonFactory = _mockJSONFactory(false);
+
+		ReflectionTestUtil.setFieldValue(
+			functionCaptchaImpl, "_jsonFactory", jsonFactory);
+
+		try {
+			functionCaptchaImpl.validateChallenge(httpServletRequest);
+
+			Assert.fail();
+		}
+		catch (Exception exception) {
+		}
+	}
+
+	private JSONFactory _mockJSONFactory(boolean success) throws Exception {
+		JSONFactory jsonFactory = Mockito.mock(JSONFactory.class);
+
+		Mockito.when(
+			jsonFactory.createJSONObject()
+		).thenReturn(
+			new JSONObjectImpl()
+		);
+
+		JSONObject jsonObject = new JSONObjectImpl();
```

Comment:
@manuelecastro, try to use the JSONUtil to build the jsonObjects if possible. Thx

Reply:
@manuelecastro After doing the final review here, I just realized that using JSONUtil is not the correct way to mock the JSONFactory since JSONUtil uses JSONFactory internally. To avoid strange issues in the future and follow other JSONFactory mocking usages, I will use your first approach.

Sorry for that :wink: